### PR TITLE
Update TableConfigCollection to add GemdQuery config generator

### PR DESCRIPTION
**Pull request by kroenlein**
# Citrine Python PR

## Description 
This is a sub-element for supporting https://github.com/CitrineInformatics/platform-backend/pull/4731.  This implements existing aspects of the API, and corrects some minor inconsistencies in the existing GemdQuery implementation.

### PR Type:
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Maintenance (non-breaking change to assist developers)

### Adherence to team decisions
- [x] I have added tests for 100% coverage
- [x] I have written Numpy-style docstrings for every method and class.
- [x] I have communicated the downstream consequences of the PR to others.
- [x] I have bumped the version in __version\__.py

### CODE REVIEW

File: src/citrine/gemd_queries/filter.py
Comment (pacdaemon) [inline src/citrine/gemd_queries/filter.py:59]: I was taking a look at the `AllIntergerFilter` case class in the backend, and it looks like it has a field called `inclusive` which defaults to `true`. Was that omission on purpose? I wonder if it is the case that it doesn't add value.

File: src/citrine/gemtables/variables.py
Comment (pacdaemon) [inline src/citrine/gemtables/variables.py:361]: Just double-checking, Here we are not mapping `attribute_template_type`, I do see it in the openapi spec, is this because we are not going to use it on the SDK?

File: src/citrine/resources/table_config.py
Comment (pacdaemon) [inline src/citrine/resources/table_config.py:555]: Nice catch

File: tests/resources/test_table_config.py
Comment (pacdaemon) [inline tests/resources/test_table_config.py:389]: Is there any code change that can be introduced to falsify this assertion? I'm just curious about why you are validating it. It might be obvious but I'm not super familiar with this.

**Review by pacdaemon — DISMISSED** [review]
I left a few comments, but other than that it looks good. If there is a need for addressing any of the concerns feel free to ping me for a second round, otherwise feel free to merge it.

File: src/citrine/gemd_queries/filter.py
Comment (kroenlein) [reply src/citrine/gemd_queries/filter.py:59]: I'd missed that.  I'll add it here as well.  🙇

File: src/citrine/gemtables/variables.py
Comment (anoto-moniz) [inline src/citrine/gemtables/variables.py:370]: Are `null` and `[]` actually treated differently by the backend? If not, I'd suggest there's no need to make this `Optional`.

File: src/citrine/gemd_queries/filter.py
Comment (anoto-moniz) [inline src/citrine/gemd_queries/filter.py:59]: You know my thoughts on this from our argument two weeks ago: changing types is an interface break. This is a mild one, as I can't think of a standard operation it would break, since Python's standard division always returns a float, and I'm unsure why a user would do something like `isinstance(value, float)`. But it nonetheless gets into the business of assuming which interface breaks the user cares about, which is a Bad Idea.

I've no intention of wasting more energy on such a petty problem, but I also won't be approving this PR as long as it's left in. Seeing as there are other approvers, and you've already got one from Pablo, it won't stand in your way.

File: src/citrine/gemtables/variables.py
Comment (kroenlein) [reply src/citrine/gemtables/variables.py:361]: It is in the openapi spec, but that field is not populated by any of the existing Variable objects in the SDK.  For the life of me, I don't see what impact that field is supposed to have, based on digging through platform-backed.

File: tests/resources/test_table_config.py
Comment (kroenlein) [reply tests/resources/test_table_config.py:389]: The table config returned by the registration route augments the table config returned by the from_query route by adding registration metadata to the object (id, version number...).  This is making sure that when you flip the `register_config` flag, you get the return of the register route, not the from_query route.

In theory, this test could be rewritten as `The table config has meta data, and if I ignore it, then they are the same` but I don't think that adds meaningful additional coverage.

File: src/citrine/gemd_queries/filter.py
Comment (kroenlein) [reply src/citrine/gemd_queries/filter.py]: I have reverted this fix, as it was tangential to the actual goal of the PR.

In this case, this is not a breaking change because the GemdQuery API is fundamentally broken and does not support any filters, so no user workflow could actually depend on this behavior.

File: src/citrine/gemtables/variables.py
Comment (kroenlein) [reply src/citrine/gemtables/variables.py:370]: I don't disagree, but every existing Variable in the SDK uses this pattern.

**Review by anoto-moniz — APPROVED** [review]

**Pull request merged**
